SIMD Divergence Optimizations (#83)
sampsyo/cs6120: content/blog/2019-10-23-divergence-optimizations/index.md

Although we target a specific architecture, every SIMD (Intel Integrated GPUs) and SIMT (nVidia and AMD Discrete GPUs) architecture can benefit from divergence optimizations.

### ISA

Predication was added to Bril to target the aforementioned hardware building upon the vector instructions added in our Project 1. A new instruction `vcmp` writes to a predicate register. A predicate register and its complement can be optionally specified before a vector instruction to mask off certain lanes of the ALU. Finally, two vector registers can be merged using the mask from a predicate register with a `vphi` instruction. An alternate implementation could remove the merge instruction and write to the same register directly at different indices. However, the former approach works better with the SSA format.

```python
# initialize vectors
  va: vector = ...;
  vb: vector = ...;

# generate the predicate
  p0: pred = vcmp va vb;

# ignore lanes based on predicate mask
  (p0)  vc0: vector = vadd va vb;
# ignore lanes based on complement of predicate mask
  (!p0) vc1: vector = vsub va vb;

# merge lanes back together
  vc: vector = vphi p0 vc0 vc1;

> 

Reviewer: Extra blank line. (This occurs a few more times below too, and it can make the rendered version look somewhat weird.)

| ------------- | ------------------- |
| Instruction is `s2v` | A different scalar value is loaded into each index of a vector register |
| Instruction is `vload`  | Unknown values are loaded into each element of a vector register |
| Any data dependency is divergent  | An incoming edge in the dataflow graph is already divergent due to one of the previous conditions |

It's possible that during runtime some vectors marked as divergent might turn out to be convergent. For example, a `vload` may load in contiguous elements with the same values. However, there is no way to optimize for this case statically.

## Divergence Optimizations

### Instruction Swapping

Once we know which instructions are divergent and which are not, we can optimize the code on an instruction-by-instruction basis. In the previous Bril example, every vector instruction is convergent. Therefore we can swap each vector instruction with a more energy-efficient scalar instruction. The optimization is shown below. 

```python
# initialize vectors -> initialize scalars
  vec0_s: int = id v0;
  vec1_s: int = id v1;

# add vectors -> add scalars
  vec2_s: int = add vec0_s vec1_s;

```

> We implement a 'swap table' that matches a vector instruction with a functionally equivalent scalar instruction. An alternate design would be to annotate each original scalar instruction with a vector length and just change the vector length instead of doing a swap. Our swap table is given below along with a description of each instruction reproduced from above.

Reviewer: ```suggestion
We implement a "swap table" that matches a vector instruction with a functionally equivalent scalar instruction. An alternate design would be to annotate each original scalar instruction with a vector length and just change the vector length instead of doing a swap. Our swap table is given below along with a description of each instruction reproduced from above.
```

|     Vector Instruction    | Description | Scalar Instruction |
| ------------- | ------------- | ------------- |
| vadd  | c[i] = a[i] + b[i]  | add |
| vsub  | c[i] = a[i] - b[i]  | sub |
| vmul  | c[i] = a[i] * b[i]  | mul |
| vdiv  | c[i] = a[i] / b[i]  | div |
| s2v  | c = (v0,v1,v2,v3)  | id |
| s2vb  | c = (v0,v0,v0,v0)  | id |
| v2s  | c = v[i]  | id |
| gather  | c[0,1,2,3] = mem[a0,a1,a2,a3]  | lw |
| scatter  | mem[a0,a1,a2,a3] = a[0,1,2,3]  | sw |
| vload  | c[i] = mem[i]  | Can't optimize |
| vstore  | mem[i] = a[i]  | Can't optimize |
| vcmp  | pred = a[i] == b[i]  | Can't optimize |
| vphi  | c[i] = pred ? a[i] : b[i]  | Can't optimize |

> Notably, we can't optimize across `vload` and `vstore` instructions because different memory addresses are always accessed. However, in certain cases a `gather` and `scatter` can access the exact same memory location if the address vector is the same for each. In this case, the access will be redundant, which will waste memory energy and potentially execution time. Even though we perform the `scatter`/`gather` optimization in the compiler, it is likely that the hardware implementation would also detect this case and avoid the redundant accesses.

Reviewer: Sounds plausible enough—but are you saying that the hardware is more likely to optimize this than, say, a single-lane vector add? It's possible, I suppose, because it could be worth the added critical-path latency. But it doesn't seem fundamentally different from the other operations you optimize.

### Vector Regeneration

An instruction swap could create a register type mismatch between the result of the optimized instruction and future vector instructions that use that result. For this reason, a second pass is added to the optimization algorithm. After the scalar instructions have been created, we traverse each instruction in program order and detect when a vector argument points to a scalar register. Upon detection, an `s2vb` instruction is generated to effectively cast a scalar value to a vector value. The faulting instruction argument is then updated to the new vector value produced by this instruction.

The benefits of replacing a vector with a scalar outweighs the additional `s2vb` instruction. For example, a vector instruction with length four consumes three more ALU ops than a scalar instruction while an additional `s2vb` only consumes a single ALU op (we assume single-op broadcast). The overall benefit is then two ALU ops worth of energy savings.

### Predication Removal

> Predicated vector instructions can also be simplified even in the case of a divergent predicate value. Every lane that is active in the vector instruction may still perform redundant work. Consider the Bril example below. The predicate `p0` is divergent because it's input `vec2` and `vec3` are divergent. However, the predicated vector instructions `vec4` and `vec5` are convergent because their inputs are convergent.

Reviewer: ```suggestion
Predicated vector instructions can also be simplified even in the case of a divergent predicate value. Every lane that is active in the vector instruction may still perform redundant work. Consider the Bril example below. The predicate `p0` is divergent because its inputs `vec2` and `vec3` are divergent. However, the predicated vector instructions `vec4` and `vec5` are convergent because their inputs are convergent.
```

#### Benchmarks

We evaluate the effectiveness of the divergence optimizations on synthetic benchmarks. We take benchmark inspiration from the examples in [these](https://ieeexplore.ieee.org/document/6494995) [papers](https://hal.inria.fr/hal-00909072v2/document). All benchmarks have a 2D loop nest. We vectorize over each outer loop and unroll over each inner loop because we do not support most control flows. We manually unroll the inner loop twice for each benchmark as it allows us to get a sense of the dynamic ALU ops without actually running the program. The number of ALU ops for the baseline and optimized version of each benchmark is shown in the table below.

|     Benchmark    | Description | Baseline Ops | Optimized Ops | Improvement (%) |
| ------------- | ------------| ------------- | ------------ | ---------- |
| FIR  | 2D FIR filter | 60 | 53 | 12
| FIR-pred | 2D FIR filter with single conditional | 87 | 81 | 7
| Synthetic | Sum of (a[outer] * b[inner]) / c[inner] | 53 | 39 | 26

> The optimization does lead to improvement in the number of ALU Ops for the listed benchmarks. It's hard to say exactly what a fair baseline would be because we don't know what a SPMD front-end would actually emit. For example, loads that aren't contiguous use `gather` in our baseline. In the case where each address is the same (i.e. indexed by the inner loop iterator), the `gather` can be turned into a `lw`. It's possible that this is obvious enough for a SPMD front-end to do automatically. We don't have a SPMD to Bril compiler, so we can't truly quantify a realistic improvement.

Reviewer: ```suggestion
The optimization does lead to improvement in the number of ALU Ops for the listed benchmarks. It's hard to say exactly what a fair baseline would be because we don't know what a SPMD front-end would actually emit. For example, loads that aren't contiguous use `gather` in our baseline. In the case where each address is the same (i.e., indexed by the inner loop iterator), the `gather` can be turned into a `lw`. It's possible that this is obvious enough for a SPMD front-end to do automatically. We don't have a SPMD to Bril compiler, so we can't truly quantify a realistic improvement.
```

## Shortcomings

These are things that we didn't do, but would have improved the implementation and empirical results.

### SSA

The code must be in SSA form to perform divergence analysis on programs with arbitrary control flow. A control dependence can be converted into a data dependence with a `phi` instruction. These data dependencies fit naturally into the dataflow algorithm used in divergence analysis.

We were not able to implement transformations to and from SSA although we did successfully implement a `phi` instruction. To work around this limitation, we manually wrote our tests and benchmarks in an SSA-like form.

### Synthetic Benchmarks

Our benchmark selection was weak for two reasons. First, as described above, we had to manually code in SSA form which made programming in Bril meticulous. The second challenge was that we could not use a high-level language to create Bril programs. The current TypeScript front-end does not support vector instructions nor the SPMD model. 

## Conclusion

> We implemented divergence analysis and optimizations based on that analysis. We focused on swapping expensive vector instructions for cheaper scalar instructions when the vector instruction did redundant work. We quantify our optimization by comparing the number of ALU Ops executed by the un-optimized baseline and optimized version. Our results show an overall reduction in ALU Ops.

Reviewer: ```suggestion
We implemented divergence analysis and optimizations based on that analysis. We focused on swapping expensive vector instructions for cheaper scalar instructions when the vector instruction did redundant work. We quantify our optimization by comparing the number of ALU ops executed by the un-optimized baseline and optimized version. Our results show an overall reduction in ALU ops.
```
